### PR TITLE
feat(medusa): Ordering products on retrieval

### PR DESCRIPTION
**Pull request by adrien2p**
**What**

Move to transformQuery which adds a default ordering and also allows to order the product list from the store API

**How**
Among other things, fix the product repo to allow ordering by either a key from the product or a key from a relation

FIXES CORE-911
FIXES CORE-901

### CODE REVIEW

**Comment by edihasaj** [comment]
This is great 🙌, thanks

**Comment by edihasaj** [comment]
Thanks @adrien2p, this is really helpful 🙌

**Comment by adrien2p** [comment]
> Thanks @adrien2p, this is really helpful 🙌

@edihasaj have you tried the pr already?

**Comment by edihasaj** [comment]
Not yet, I will try it today most probably, and let you know 👍

File: packages/medusa/src/repositories/product.ts [hunk 1]
```diff
@@ -298,7 +300,10 @@ export class ProductRepository extends Repository<Product> {
     const entitiesIdsWithRelations = await this.queryProductsWithIds(
       entitiesIds,
       groupedRelations,
-      withDeleted
+      withDeleted,
+      Array.isArray(idsOrOptionsWithoutRelations)
```
Comment (pKorsholm) [inline packages/medusa/src/repositories/product.ts:304]: since you've added another param (order) to queryProductsWithIds, I think you need to add an empty array as selector as that was omitted due to being an optional parameter

**Review by pKorsholm — COMMENTED** [review]
Just querying with the order property seems to be working nicely. I do have a question though

Question:  When you add the param `q` ordering doesn't seem to have an effect anymore. Is this by design (or have I messed up something locally)?
It would seem to me that you would want to be able to search and order at the same time. Though I also recognize the reasoning for this pr is that the repositories are very convoluted atm and could use some love, and that might have a hand in play regarding ordering and querying at the same time.

File: integration-tests/api/__tests__/store/__snapshots__/products.js.snap [hunk 2]
```diff
@@ -1,450 +0,0 @@
-// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (srindom) [inline integration-tests/api/__tests__/store/__snapshots__/products.js.snap:1]: question: why is this deleted?

File: packages/medusa/src/repositories/product.ts (hunk 1 above)
Comment (adrien2p) [reply packages/medusa/src/repositories/product.ts:304]: Nice catch, regarding the free text search I did not touched it. I ll add it as part of that pr

Comment (adrien2p) [reply packages/medusa/src/repositories/product.ts:304]: I also tried to cleaned up a bit the repo while not make it more complicated :D

File: integration-tests/api/__tests__/store/__snapshots__/products.js.snap (hunk 2 above)
Comment (adrien2p) [reply integration-tests/api/__tests__/store/__snapshots__/products.js.snap:1]: Because it has been replaced by the expect instead. As we agreed some times ago that the snap will disapeared

File: .changeset/bright-moose-draw.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@medusajs/medusa": patch
+---
+
+feat: Allow store product list to be ordered (as well as by default)
```
Comment (olivermrbl) [inline .changeset/bright-moose-draw.md:5]:
```suggestion
feat(medusa): Order products on retrieval
```

File: packages/medusa/src/repositories/product.ts [hunk 4]
```diff
@@ -58,47 +61,36 @@ export class ProductRepository extends Repository<Product> {
...
-      })
+      qb.leftJoin(`${productAlias}.tags`, "tags").andWhere(
+        `tags.id IN (:...tag_ids)`,
+        {
+          tag_ids: tags.value,
+        }
+      )
     }
 
+    const arePricesJoined = !!price_lists
```
Comment (olivermrbl) [inline packages/medusa/src/repositories/product.ts:82]: suggestion(non-blocking): Move closer to its usage e.g. line 108

Comment (olivermrbl) [reply packages/medusa/src/repositories/product.ts:82]: suggestion(non-blocking): Rename to `joinWithPriceLists`

File: packages/medusa/src/utils/get-query-config.ts [hunk 5]
```diff
@@ -66,8 +66,10 @@ export function getListConfig<TModel extends BaseEntity>(
     expandFields = expand
   }
 
-  const orderBy: Record<string, "DESC" | "ASC"> = order ?? {
-    created_at: "DESC",
+  const orderBy = order
+
+  if (!Object.keys(order).length) {
+    orderBy["created_at"] = "DESC"
```
Comment (olivermrbl) [inline packages/medusa/src/utils/get-query-config.ts:72]: praise: nice!

File: packages/medusa/src/utils/repository.ts [hunk 6]
```diff
@@ -163,3 +165,66 @@ export function mergeEntitiesWithRelations<T>(
...
+              (relationMetadata) => relationMetadata.propertyName === child
+            )
+            const shouldApplyJoin = shouldJoin(child)
+            if (shouldApplyJoin) {
+              qb.leftJoin(`${parent}.${relationMetadata!.propertyPath}`, child)
+            }
+            return [child, relationMetadata!.inverseEntityMetadata]
+          },
+          [alias, repository.metadata]
+        )
```
Comment (olivermrbl) [inline packages/medusa/src/utils/repository.ts:211]: suggestion(non-blocking): Though this is super nice, it is also pretty hard to read. I think we should add a couple of comments in case we need to revisit it.

Comment (olivermrbl) [reply packages/medusa/src/utils/repository.ts:211]: We could include an example of parsing `order` in the JSDoc?

File: integration-tests/api/__tests__/store/__snapshots__/products.js.snap (hunk 2 above)
Comment (srindom) [reply integration-tests/api/__tests__/store/__snapshots__/products.js.snap:1]: Sorry wasn't aware :)

Comment (adrien2p) [reply integration-tests/api/__tests__/store/__snapshots__/products.js.snap:1]: ahah, I think it started here https://medusacorp.slack.com/archives/C025KMS13SA/p1668178263149179

**Review by srindom — APPROVED** [review]
LGTM! Nice work
